internal/runtime/runtime.go

```diff
@@ -93,6 +102,30 @@ func (r *Runtime) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 		helpers.RespondHTTP(helpers.Response{StatusCode: http.StatusInternalServerError}, err, resp)
 		return
 	}
-	response, err := r.Handler.Process(body, headers)
-	helpers.RespondHTTP(response, err, resp)
+	result, err := r.Handler.Process(body, headers)
+	// Extensions
+	r.extensions(result, err)
+	helpers.RespondHTTP(result.Response, err, resp)
+}
+
+// extensions is a helper function to execute additional runtime extensions
+func (r *Runtime) extensions(promotionResult *promotion.Result, err error) {
+	// send feedback commit status: error
+	if err != nil {
+		status := controllers.CommitStatusFailure
+		var promotionErr *promotion.InternalError
+		if errors.As(err, &promotionErr) {
+			status = controllers.CommitStatusError
+		}
+		if statusErr := r.Handler.SendFeedbackCommitStatus(promotionResult, err, status); statusErr != nil {
+			r.logger.Error("failed to send feedback commit status", slog.Any("error", statusErr))
+		}
+	}
+
+	// Rate limits: if fetchRateLimits is enabled, fetch rate limits
+	if rateLimits, err := r.Handler.RateLimits(promotionResult); err != nil {
```

Comment:
We need to be smart here: this should only return the rate limits from the most recent API call, and definitely should *not* call `/rate_limits` in the case of invalid requests.